File: lys/가장 큰 수.js

function solution(numbers) {

    let answer = numbers.sort(function(a,b) { 
        a = a.toString();
        b = b.toString();
        return (b+a) - (a+b);
    }).join('');

    return answer[0] === '0' ? '0' : answer   // 모든 숫자가 0인 경우 000 이 나오지 않게 0을 출력
}
//console.log(solution([6, 10, 2]));
//console.log(solution([0, 0, 0]));
//console.log(solution([3, 30, 34, 5, 9]));

// 문자열을 그대로 연결한 수(b+a) - 바꿔 연결한 수(a+b)가 양수이면 b를 a앞에 둠, 음수면 반대로 둠
// ex) b(3) + a(30) - a(30) + b(3) => 330 - 303 = 양수
// 3 30 순서 그대로

File: ksh/week1/가장_큰_수.java
package week1;

import java.util.List;
import java.util.ArrayList;
import java.util.Collections;

class 가장_큰_수 {
    public String solution(int[] numbers) {
        List<String> stringList = new ArrayList<>();
        for(int num : numbers){
            stringList.add(String.valueOf(num));
        }

        // 1. 사전순 내림차순 정렬
        Collections.sort(stringList, Collections.reverseOrder());
        // 람다식 숙지 필요
        Collections.sort(StringList, (a, b) -> (b + a).compareTo(a + b));

        StringBuilder sb = new StringBuilder();
        for(String str : stringList){
            sb.append(str);
        }

        // 정답을 본 부분
        if(sb.charAt(0) == '0') {
            return "0";
            // 예외 처리: 가장 큰 수가 0으로 시작하는 경우
            // 모든 원소가 0인 경우에도 "000..."이 아닌 "0"으로 반환
        }

        return sb.toString();
    }
}